Dashboard: Python 3 rebase and QA

ross-spencer opened the pull request:
QA branch  for the Dashboard Py3 migration... 

Connected to archivematica/issues#810
Replaces https://github.com/artefactual/archivematica/pull/1478 

**CR notes**

* This branch rebases all of Helen's work. 
* I made CR-based changes here https://github.com/artefactual/archivematica/pull/1575/commits/36e8f9a36ff94395ad7858879a9c09c7a3c7292e most of these are just small aesthetic improvements, catching small issues introduced by modernize, or adding recommended code, e.g. `absolute_import` mainly. 
* I borrowed the configuration for the Jenkins `matrix-qa` job to run against the most recent commit in this PR: https://github.com/artefactual/archivematica/pull/1575/commits/63c5cee09f362723dad52e43bc7a2b4fa9126afd.

E.g. 

![image](https://user-images.githubusercontent.com/1880412/75121650-fc211900-5663-11ea-997b-c1b2eea83bc9.png)

And the results look good:

![image](https://user-images.githubusercontent.com/1880412/75121652-02af9080-5664-11ea-94f1-5e6ba2409077.png)

As well as running various different pathways manually by hand when looking through the CR, e.g. I wanted to make sure new  features like the storage service report pages worked as anticipated, or to see if handling of things like processing configs changed etc. 

Additional notes:

* AMAUAT works. 
* Unit tests work.
* Migrations work as we'd expect to be able to run the tests. 
* API functions work for the range of API calls being used by black-box testing. 
* Gearman has been reverted to Gearman 2 related to the notes [here[(https://github.com/archivematica/Issues/issues/845#issuecomment-583903168).

So I believe this is looking good, it does need @artefactual/amcr approval as a new PR to merge, so hopefully these notes help and the commit layout post Helen's work.

helenst committed:
Merge Python 3 compatible changes

These are Python 2 and 3 compatible cherry picked commits from our
Python 3 sprint during August 2019.

ross-spencer commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -105,6 +107,14 @@ def strToUnicode(string, obstinate=False):
+
+
+def b64decode_string(data):
+    return base64.b64decode(data.encode("utf8")).decode("utf8")
```
NB. new helper functions to help across the Dashboard.

ross-spencer commented on src/dashboard/src/components/administration/forms.py:
```diff
@@ -468,4 +469,4 @@ def add_choice(
     def save(self, config_path):
         with open(config_path, "w") as f:
-            f.write(etree.tostring(self.xml, pretty_print=True))
+            f.write(etree.tostring(self.xml, pretty_print=True, encoding="unicode"))
```
NB. Because we're serializing to disk I think this needs to be a byte encoding `UTF-8` per https://lxml.de/parsing.html

ross-spencer commented on src/dashboard/src/main/management/commands/__init__.py:
```diff
@@ -18,9 +19,9 @@ def info(self, message):
 def boolean_input(question, default=None):
     question += '\n\nType "yes" to continue, or "no" to cancel: '
-    result = input("%s " % question)
+    result = eval(input("%s " % question))
```
NB. Check what we did in storage service here, this might not work...

replaceafill commented on src/dashboard/src/requirements/base.in:
```diff
@@ -12,10 +12,11 @@ logutils==0.3.3
 django-tastypie==0.13.2
 django-extensions==1.1.1
 elasticsearch>=6.0.0,<7.0.0
+futures==3.2.0  # used by gunicorn's async workers
```
Is this line redundant given line 19 below additionally restricts `< 3.0`?

replaceafill commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -134,12 +134,13 @@
     import configparser
 except ImportError:
-    import ConfigParser as configparser
+    import six.moves.configparser as configparser
```
Looking at how other modules import `configparser` from `six` now, I think we can remove this `ImportError` block.

ross-spencer replied:
Strange. Wonder what the specific reason for this one was. :man_shrugging:

replaceafill commented on src/dashboard/src/main/migrations/0066_archivesspace_base_url.py:
```diff
@@ -5,14 +5,15 @@
 try:
-    from urlparse import urlparse
+    from six.moves.urllib.parse import urlparse
 except ImportError:
```
I think we can remove this `ImportError` block now and just depend on the `six` import.

replaceafill commented on src/dashboard/tests/test_api_v2beta.py:
```diff
@@ -42,8 +43,8 @@ class TestAPIv2(TestCase):
             }
-        )
-    )
+        ).encode("utf8")
```
Does this not apply for the new `archivematicaFunctions.b64encode_string` function?

ross-spencer replied:
I'm not clear on what you mean here @replaceafill, can you elaborate at all?

replaceafill replied:
Yeah. Sorry about not being clear.

I noticed the string resulting from `"{location_uuid}:{relative_path}".format(...)` is then encoded as `utf8`, then encoded as `base64` and finally decoded as `utf8`. That's what the new `archivematicaFunctions.b64encode_string` helper seems to do, right? Unless I'm misinterpreting it in which case consider this noise :blush: 

 So I was thinking something along these lines:

```
path = archivematicaFunctions.b64encode_string(
    "{location_uuid}:{relative_path}".format(
        **{
               "location_uuid": "671643e1-5bec-4a5f-b244-abb76fedb0c4",
               "relative_path": "foo/bar.jpg",
        }
    )
)
```

replaceafill commented on src/MCPClient/requirements/base.txt:
```diff
@@ -27,4 +27,4 @@ unidecode==0.04.19
 
 # The following packages are considered to be unsafe in a requirements file:
-# setuptools
+# setuptools==41.0.1        # via opf-fido
```
I noticed you removed these from the Dashboard requirements in https://github.com/artefactual/archivematica/commit/36e8f9a36ff94395ad7858879a9c09c7a3c7292e. Should we remove these in the client and server as well?

ross-spencer replied:
I think the correct approach is to only modify the `.in` file, so prior to merging I'll run the compile requirements job again, and then I think that will shake this out. What might have happened in that commit was an artifact of the same process.

replaceafill reviewed (commented):
This looks good to me overall. I left a few comments with minor (potential) cleanups and have two questions:

1. Do you think we should follow the same approach when we merged the Python 3 changes for the Storage Service and address (some of) the warnings produced by `pylint`?

```bash
$ venv3/bin/pylint --py3k $(find * | grep py$)
************* Module src.components.transfer.views
src/components/transfer/views.py:210:15: W1645: Exception.message removed in Python 3 (exception-message-attribute)
src/components/transfer/views.py:215:22: W1645: Exception.message removed in Python 3 (exception-message-attribute)
src/components/transfer/views.py:239:30: W1645: Exception.message removed in Python 3 (exception-message-attribute)
************* Module src.components.administration.views
src/components/administration/views.py:216:55: W1655: dict.keys referenced when not iterating (dict-keys-not-iterating)
************* Module src.components.administration.tests.test_storage
src/components/administration/tests/test_storage.py:1:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.components.administration.tests.test_usage
src/components/administration/tests/test_usage.py:1:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.components.accounts.apps
src/components/accounts/apps.py:1:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.components.accounts.signals
src/components/accounts/signals.py:1:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.components.archival_storage.tests.test_archival_storage
src/components/archival_storage/tests/test_archival_storage.py:22:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.fpr.migrations.0031_update_normalization_rules
src/fpr/migrations/0031_update_normalization_rules.py:7:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.fpr.migrations.0032_pronom_96
src/fpr/migrations/0032_pronom_96.py:4:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module src.fpr.migrations.0033_update_idtools
src/fpr/migrations/0033_update_idtools.py:5:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module test_helpers
tests/test_helpers.py:5:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module test_middleware
tests/test_middleware.py:1:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
************* Module test_filesystem_ajax
tests/test_filesystem_ajax.py:266:42: W1646: non-text encoding used in str.decode (invalid-str-codec)
************* Module test_api
tests/test_api.py:567:42: W1646: non-text encoding used in str.decode (invalid-str-codec)

-----------------------------------
Your code has been rated at 9.99/10
```

2. I noticed the dashboard is the only component using `six==1.12.0` while the rest use `1.13.0`. Maybe it's not a big deal but I was wondering if it could be quickly updated by running `pip-compile` again?

sevein commented on src/archivematicaCommon/lib/appconfig.py:
```diff
@@ -88,8 +92,8 @@ def get_from_opts_list(self, attr, attr_opts_list, default=None):
-                ConfigParser.NoSectionError,
-                ConfigParser.NoOptionError,
+                six.moves.configparser.NoSectionError,
+                six.moves.configparser.NoOptionError,
```
This feels weird, can we bind the import to `configparser` with `import as`?

sevein commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -24,6 +24,8 @@
 """
 
 from __future__ import print_function
+from __future__ import absolute_import
```
We usually group imports from the same module.

sevein commented on src/archivematicaCommon/lib/appconfig.py:
```diff
@@ -13,7 +13,8 @@
 """
-import ConfigParser
+from __future__ import absolute_import
+import six.moves.configparser
```
The six import should go to the 3rd-party import group.

sevein commented on src/archivematicaCommon/lib/storageService.py:
```diff
@@ -1,16 +1,19 @@
+from requests.auth import AuthBase
+import six.moves.urllib.error
+import six.moves.urllib.parse
+import six.moves.urllib.request
```
This feels very Javaish! Another exmaple where we could just import `urllib` via local binding.

sevein reviewed (changes requested):
This is a lot of work, thank you! I've found some inconsistencies in the way we're doing imports and shared some examples. Can they be fixed this time around? I don't think that's something that we could do mechanically with isort.

ross-spencer commented:
I've placed this into re-review. AMAUAT look to be passing locally, but have focused mainly on `mo-aip-reingest`, `transfer-policy-check`, `aip-encrypt-mirror` plus `black-box`. I've done some light QA in the user-interface as well.

ross-spencer commented:
@replaceafill on `pylint --py3k`. My initial instinct was not to add too much to Helen's PR, but I think it was a good call. I've addressed most of those that I felt were safest for me to do this time around, and we should be better off. There are a few other `pylint` warnings in AM Common I think we can afford to look at with the other components, I think they might affect mcp-client or mcp-server more too. Thank you for the push! 

![image](https://media1.giphy.com/media/LfCt1sR1VweBy/giphy.gif?cid=790b7611234447d245d33fd513fa28853ea8e7c1b760e559&rid=giphy.gif)

replaceafill commented:
Thank you for the hard work @ross-spencer! This looks good to me :+1: I'll wait for @sevein to take another look and approve.

sevein commented on src/archivematicaCommon/lib/appconfig.py:
```diff
@@ -13,10 +13,12 @@
 
 from django.core.exceptions import ImproperlyConfigured
 
+import six.moves.configparser as ConfigParser
```
Nit: in Py3 it's named `configparser` - action not really required here but it'd be nice.

sevein commented on src/dashboard/src/settings/base.py:
```diff
@@ -16,14 +18,16 @@
 
 from django.utils.translation import ugettext_lazy as _
+from django.utils.six import StringIO
+from six import StringIO
```
I think that you can remove L29 and that'd be the last import using django.utils.six in the Dashboard!

sevein commented on src/dashboard/src/contrib/mcp/client.py:
```diff
@@ -112,7 +114,7 @@ def _rpc_sync_call(self, ability, data=None, timeout=INFLIGHT_POLL_TIMEOUT):
         response = client.submit_job(
             ability,
-            cPickle.dumps(data),
+            six.moves.cPickle.dumps(data, protocol=0),
```
It'd be nice to bind the import to `pickle` since that's the name used in Py3.

sevein commented on src/dashboard/src/components/transfer/views.py:
```diff
@@ -234,7 +231,7 @@ def cleanup_metadata_set(request, set_uuid):
-    except Exception as e:
-        response["message"] = e.message
+    except Exception as err:
+        response["message"] = "{}".format(err)
```
Being a bit pedantic here, if we didn't want to use `format` would the following be more correct?

```suggestion
        response["message"] = six.text_type(err)
```

sevein commented on src/dashboard/src/components/ingest/views.py:
```diff
@@ -38,10 +35,9 @@
 from django.utils.text import slugify
 from django.utils.translation import ugettext as _
 from django.views.generic import View
+import six.moves.cPickle
```
Bind import to `pickle`?

sevein commented on src/dashboard/src/components/helpers.py:
```diff
@@ -33,11 +33,16 @@
 from tastypie.models import ApiKey
+from six.moves.urllib.parse import urlencode, urljoin
+from six.moves import range
+from six.moves import zip
```
Group imports.

sevein commented on src/archivematicaCommon/lib/env_configparser.py:
```diff
@@ -1,7 +1,11 @@
-import ConfigParser
 import functools
 
+import six.moves.configparser as ConfigParser
```
Bind as `configparser`.

sevein reviewed (approved):
Thanks! A few more comments on import style but :+1:.

ross-spencer commented:
Thanks @sevein I've removed the excess import and changed the `"{}".format()` call to `six.text_type` as it looks cleaner. I haven't bound these final imports differently this time. My personal feeling is that changing what the binding is pointing at is nice and easy, it's a reference. Going into the body of the code to change `cPickle` to `pickle` as an example, just at this stage, is just another code-change/copy-replace too much. I'd like to link this issue back to the gearman issue as I think we can think about addressing that as we touch each of those modules again for that work.
